### PR TITLE
[SofaBaseTopology] Fix Crash when loading a vtk file generated by Gmsh using TetrahedronSetTopologyContainer as container

### PR DESCRIPTION
**Pull request by olivier-goury**
Good afternoon!
This PR proposes a fix for a crash when loading a vtk file that was generated by Gmsh.

`
 myNode.createObject('MeshVTKLoader', name='loader', filename='gmshMesh.vtk')
 myNode.createObject('TetrahedronSetTopologyContainer', src="@loader")
`

The mesh generated by default with Gmsh gives a list of edges and triangles that is incomplete and contains only edges and triangles on the boundary of the mesh. However the list of Tetrahedra is complete. This leads to a crash when using a TetrahedronSetTopologyContainer.
This PR tries to address this problem.

Cheers!
Olivier

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by epernod** [comment]
Hi,
not sure to understand, are the edges in the VTK correct regarding the tetra? Or do they correspond to the border?

**Comment by olivier-goury** [comment]
> Hi,
> not sure to understand, are the edges in the VTK correct regarding the tetra? Or do they correspond to the border?

There are not consistent with the tetra. Basically, the tetra information contains all the tetras as you would expect, but the edges and triangles correspond only to the boundary of the mesh.

**Comment by epernod** [comment]
Ok and you want to keep them or you want to recompute them? 
It is not easy to read the fixin the PR.

I made some change recently in the container to be sure that all info edges/triangles and tetrahedra are coherent.
If you want to create the border topology you should use TopologicalMapping. Or link data by data to different containers and not using src.

Otherwise to recompute the sub element, you should use a "hidden" option in the VTKLoader: See what has been done for Gmsh in PR #986

**Comment by olivier-goury** [comment]
Hi,
I don't actually want to use them, it's just that it was crashing! I created the mesh in Gmsh but exported it in vtk. Exporting it in the msh format does not lead to a crash indeed.
I can understand that my fix is not super clear!  
Essentially, in the code when it starts using the edges and something goes wrong, I move on to the option like if no edges had been given (For example in the function createEdgesInTetrahedronArray() with the if (hasEdges) ). I hope it clarifies a little bit.

File: SofaKernel/modules/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp [hunk 1]
```diff
@@ -234,12 +199,50 @@ void TetrahedronSetTopologyContainer::createEdgesInTetrahedronArray()
                         foundEdge=true;
                     }
                 }
-
+                if (foundEdge == false)  // The edges and triangles in this mesh are only on the border: this mesh was probably created using Gmsh
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp]: and what happen if it enter this case? 
The container will stay without edges?

File: SofaKernel/modules/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp [hunk 2]
```diff
@@ -375,17 +386,16 @@ void TetrahedronSetTopologyContainer::createTetrahedraAroundTriangleArray ()
...
+    if (m_trianglesInTetrahedron.size() !=0){
     for (size_t i=0; i<getNumberOfTetrahedra(); ++i)
     {
         // adding tetrahedron i in the shell of all neighbors triangles
         for (TriangleID j=0; j<4; ++j)
         {
             m_tetrahedraAroundTriangle[ m_trianglesInTetrahedron[i][j] ].push_back( (TetrahedronID)i );
         }
-    }
+    }}
```
Comment (guparan) [inline SofaKernel/modules/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp]: This block still needs some indentation

File: SofaKernel/modules/SofaBaseTopology/TriangleSetTopologyContainer.cpp [hunk 3]
```diff
@@ -292,19 +259,19 @@ void TriangleSetTopologyContainer::createEdgesInTriangleArray()
...
 
         for (size_t edge=0; edge<numEdges; ++edge)  //Todo: check if not better using multimap <PointID ,TriangleID> and for each edge, push each triangle present in both shell
         {
             edgesAroundVertexMap.insert(std::pair<PointID, EdgeID> (m_edge[edge][0], (EdgeID)edge));
             edgesAroundVertexMap.insert(std::pair<PointID, EdgeID> (m_edge[edge][1], (EdgeID)edge));
         }
-
-        for(size_t i=0; i<numTriangles; ++i)
+        size_t i=0;
+        while(i<numTriangles && foundEdge==true)
```
Comment (guparan) [inline SofaKernel/modules/SofaBaseTopology/TriangleSetTopologyContainer.cpp]: `for ( size_t i = 0 ; (i < numTriangles) &&  (foundEdge == true) ; ++i )` ?

File: SofaKernel/modules/SofaBaseTopology/TriangleSetTopologyContainer.cpp [hunk 4]
```diff
@@ -292,19 +259,19 @@ void TriangleSetTopologyContainer::createEdgesInTriangleArray()
...
         {
             const Triangle &t = m_triangle[i];
             // adding edge i in the edge shell of both points
-            for(unsigned int j=0; j<3; ++j)
+            int j=0;
+            while(j<3 && foundEdge==true)
```
Comment (guparan) [inline SofaKernel/modules/SofaBaseTopology/TriangleSetTopologyContainer.cpp]: `for ( unsigned int j = 0 ; (j < 3) && (foundEdge == true) ; ++j )` ?

File: SofaKernel/modules/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp [hunk 5]
```diff
@@ -207,20 +174,21 @@ void TetrahedronSetTopologyContainer::createEdgesInTetrahedronArray()
...
 
         for (EdgeID edge=0; edge<numEdges; ++edge)  //Todo: check if not better using multimap <PointID ,TriangleID> and for each edge, push each triangle present in both shell
         {
             edgesAroundVertexMap.insert(std::pair<PointID, EdgeID> (m_edge[edge][0],edge));
             edgesAroundVertexMap.insert(std::pair<PointID, EdgeID> (m_edge[edge][1],edge));
         }
-
-        for(size_t i=0; i<numTetra; ++i)
+        size_t i=0;
+        while(i<numTetra && foundEdge == true)
```
Comment (guparan) [inline SofaKernel/modules/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp]: `for ( size_t i = 0 ; (i < numTetra) && (foundEdge == true) ; ++i )` ?

File: SofaKernel/modules/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp [hunk 6]
```diff
@@ -207,20 +174,21 @@ void TetrahedronSetTopologyContainer::createEdgesInTetrahedronArray()
...
         {
             const Tetrahedron &t = m_tetrahedron[i];
             // adding edge i in the edge shell of both points
-            for(EdgeID j=0; j<6; ++j)
+            EdgeID j=0;
+            while(j<6 && foundEdge == true)
```
Comment (guparan) [inline SofaKernel/modules/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp]: `for ( EdgeID j = 0 ; (j < 6) && (foundEdge == true) ; ++j )` ?

**Review by guparan — CHANGES_REQUESTED** [review]

**Comment by olivier-goury** [comment]
Thanks for your comments Guillaume, I agree with your changes!

**Review by guparan — APPROVED** [review]

**Review by epernod — APPROVED** [review]

**Comment by epernod** [comment]
I'm ok with the change regarding the edges. I can't see some border effect right now... and as all the non regression tests are green I'm 99% confident!
I just added some additional warning and tests inside the tetrahedron methods you fixed to match what has been done inside the TriangleSetTopologyContainer. You could double check if needed.
Thanks, I missed those one.
:+1:

**Pull request merged**